key.go

```diff
@@ -0,0 +1,806 @@
+// int or tstr).
+func (kt *KeyType) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key type value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := KeyTypeFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*kt = v
+	} else {
+		v := raw.Int()
+
+		if v == 0 {
+			// 0  is reserved, and so can never be valid
+			return fmt.Errorf("invalid key type value 0")
+		}
+
+		if v > 4 || v < 0 || v == 3 {
+			return fmt.Errorf("unknown key type value %d", v)
+		}
+
+		*kt = KeyType(v)
+	}
+
+	return nil
+}
+
+const (
+
+	// Invalid/unrecognised curve
+	CurveInvalid Curve = 0
+
+	// NIST P-256 also known as secp256r1
+	CurveP256 Curve = 1
+
+	// NIST P-384 also known as secp384r1
+	CurveP384 Curve = 2
+
+	// NIST P-521 also known as secp521r1
+	CurveP521 Curve = 3
+
+	// X25519 for use w/ ECDH only
+	CurveX25519 Curve = 4
+
+	// X448 for use w/ ECDH only
+	CurveX448 Curve = 5
+
+	// Ed25519 for use /w EdDSA only
+	CurveEd25519 Curve = 6
+
+	// Ed448 for use /w EdDSA only
+	CurveEd448 Curve = 7
+)
+
+// Curve represents the EC2/OKP key's curve. See:
+// https://datatracker.ietf.org/doc/html/rfc8152#section-13.1
+type Curve int64
+
+func CurveFromString(v string) (Curve, error) {
+	switch v {
+	case "P-256":
+		return CurveP256, nil
+	case "P-384":
+		return CurveP384, nil
+	case "P-521":
+		return CurveP521, nil
+	case "X25519":
+		return CurveX25519, nil
+	case "X448":
+		return CurveX448, nil
+	case "Ed25519":
+		return CurveEd25519, nil
+	case "Ed448":
+		return CurveEd448, nil
+	default:
+		return CurveInvalid, fmt.Errorf("unknown curve value %q", v)
+	}
+}
+
+func (c Curve) String() string {
+	switch c {
+	case CurveP256:
+		return "P-256"
+	case CurveP384:
+		return "P-384"
+	case CurveP521:
+		return "P-521"
+	case CurveX25519:
+		return "X25519"
+	case CurveX448:
+		return "X448"
+	case CurveEd25519:
+		return "Ed25519"
+	case CurveEd448:
+		return "Ed448"
+	default:
+		return "unknown curve value " + strconv.Itoa(int(c))
+	}
+}
+
+// MarshalCBOR marshals the KeyType as a CBOR int.
+func (c Curve) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(c))
+}
+
+// UnmarshalCBOR populates the KeyType from the provided CBOR value (must be
+// int or tstr).
+func (c *Curve) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid curve value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := CurveFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*c = v
+	} else {
+		v := raw.Int()
+
+		if v < 1 || v > 7 {
+			return fmt.Errorf("unknown curve value %d", v)
+		}
+
+		*c = Curve(v)
+	}
+
+	return nil
+}
+
+// Key represents a COSE_Key structure, as defined by RFC8152.
+// Note: currently, this does NOT support RFC8230 (RSA algorithms).
+type Key struct {
+	keyStruct
```

Comment:
Embedding a private struct with public properties is normally not a good idea, as those properties are not listed by `go doc`, therefore they won't be documented in pkg.go.dev.

I would suggest moving all `keyStruct` properties to `Key` and tag them as `cbor:-` so they are ignored when encoding a key as cbor. Then remove `keyStruct`.